### PR TITLE
Adds a column argument to unique and nonunique

### PR DESCRIPTION
**Pull request by mkborregaard**
Adds a second optional argument to unique(), nonunique() and unique!()
that specifies which columns to base the comparison on

### CODE REVIEW

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -587,22 +590,40 @@ function nonunique(df::AbstractDataFrame)
     res
 end
 
+function nonunique(df::AbstractDataFrame, columns::Any)
+    nonunique(df[columns])
+end
+
+function nonunique(df::AbstractDataFrame, columns::Int)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Why did you add these two redundant methods?

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -587,22 +590,40 @@ function nonunique(df::AbstractDataFrame)
     res
 end
 
+function nonunique(df::AbstractDataFrame, columns::Any)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Could be written as a one liner.

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -551,11 +551,13 @@ Indexes of complete cases (rows without NA's)
 
 ```julia
 nonunique(df::AbstractDataFrame)
+nonunique(df::AbstractDataFrame, columns)
 ```
 
 ### Arguments
 
 * `df` : the AbstractDataFrame
+* `columns` : specifies the columns to compare
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: "specifies" isn't useful here. OTOH, it would be useful to precise that columns can be either a symbol, an integer, or vectors of these types.

**Comment by nalimilan** [comment]
Thanks. Could you also add tests for the new functions?

File: src/abstractdataframe/abstractdataframe.jl (hunk 1 above)
Comment (mkborregaard) [inline src/abstractdataframe/abstractdataframe.jl]: Because otherwise `unique(df, 2)` fails (must be `unique([2])` which I find less intuitive). I could do `Union{Int, Symbol} ` if that is preferred?

**Comment by mkborregaard** [comment]
I will fix these things and add the tests tomorrow.

File: src/abstractdataframe/abstractdataframe.jl (hunk 3 above)
Comment (kmsquire) [inline src/abstractdataframe/abstractdataframe.jl]: Since `columns` is also a function name, might want to rename this to `cols` (which, e.g., is the name used in `sort`).

File: src/abstractdataframe/abstractdataframe.jl (hunk 1 above)
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Ah, indeed. Using `Union` is better IMHO. Also, a one-liner would look nicer.

Comment (alyst) [inline src/abstractdataframe/abstractdataframe.jl]: I would also use `col` as the argument name.

File: src/abstractdataframe/abstractdataframe.jl (hunk 3 above)
Comment (mkborregaard) [inline src/abstractdataframe/abstractdataframe.jl]: I have changed that in the new version

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -587,22 +590,30 @@ function nonunique(df::AbstractDataFrame)
     res
 end
 
+nonunique(df::AbstractDataFrame, cols::Any) = nonunique(df[[cols]])
+nonunique{T <: Any}(df::AbstractDataFrame, cols::AbstractVector{T}) = nonunique(df[cols])
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: You don't need `{T <: Any}` nor `{T}` here.

File: test/data.jl [hunk 5]
```diff
@@ -305,6 +305,20 @@ module TestData
     # @test (3 in unique(dv))
     # @test (NA in unique(dv))
 
+    # test nonunique() with extra argument
+    df = DataFrame(a = ["a", "b", "a", "b", "a", "b"], b = 1:6, c = [1:3;1:3])
+    df = vcat(df, df)
+    @test isequal(sum(nonunique(df)), 6)
```
Comment (nalimilan) [inline test/data.jl]: Use `==` instead of `isequal` unless there's a specific reason not to (e.g. `NaN`).

Could you test the actual contents of the data frame instead? That way, we're completely sure the code will continue to work, and not return garbage.

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (alyst) [inline src/abstractdataframe/abstractdataframe.jl]: I think the default method (i.e. `cols::Any`) should be the "vector" one, and for the "scalar" case the parameters should be explicitly listed (`col::Union{Real,Symbol}` as in DataFrame's `ColumnIndex`). Otherwise `Colon()` could be incorrectly dispatched.

**Comment by nalimilan** [comment]
Tests fail on 0.3. Should we start requiring 0.4, or use `@compat`?

**Comment by nalimilan** [comment]
I guess we may as well drop support for Julia 0.3, since https://github.com/JuliaStats/DataFrames.jl/pull/921 will require StatsBase 0.8.0, which doesn't run there either.

**Comment by mkborregaard** [comment]
The PR should be ready to merge, IMO.

**Comment by nalimilan** [comment]
There's still a failure in the tests due to `repeat`.

File: test/data.jl [hunk 6]
```diff
@@ -305,6 +305,24 @@ module TestData
...
     # @test (NA in unique(dv))
 
+    # test nonunique() with extra argument
+    df1 = DataFrame(a = ["a", "b", "a", "b", "a", "b"], b = 1:6, c = [1:3;1:3])
+    df = vcat(df1, df1)
+    @test nonunique(df) == [rep(false,6); rep(true, 6)]
+    @test nonunique(df, :) == [rep(false,6); rep(true, 6)]
+    @test nonunique(df, Colon()) == [rep(false,6); rep(true, 6)]
+    @test nonunique(df, :a) == [rep(false,2); rep(true, 10)]
+    @test nonunique(df, [1, 3]) == [rep(false,6); rep(true, 6)]
```
Comment (nalimilan) [inline test/data.jl]: Could you add a test with a single integer, and one with a vector of symbols?

File: test/data.jl [hunk 7]
```diff
@@ -305,6 +305,24 @@ module TestData
...
+
+    # Test unique() with extra argument
+    @test unique(df) == df1
+    @test unique(df, :) == df1
+    @test unique(df, Colon()) == df1
+    @test unique(df, 2:3) == df1
+    @test unique(df, 3) == df1[1:3,:]
+    @test unique(df, [:a, :c]) == df1
```
Comment (nalimilan) [inline test/data.jl]: Same here: please add a test with a vector of integers, and with a single symbol.

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -552,11 +552,13 @@ Indexes of complete cases (rows without NA's)
 
 ```julia
 nonunique(df::AbstractDataFrame)
+nonunique(df::AbstractDataFrame, cols)
 ```
 
 **Arguments**
 
 * `df` : the AbstractDataFrame
+* `cols` : the column(s) to compare – any valid DataFrames column specifier will work
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: `aggregate` uses the following description for this argument:
```
* `cols` : a column indicator (Symbol, Int, Vector{Symbol}, etc.)
```

**Comment by mkborregaard** [comment]
Thanks, should be fixed now.

**Comment by nalimilan** [comment]
Thanks! I'll leave the PR open for a while to leave a change for others to comment before merging.

**Comment by alyst** [comment]
LGTM as well.
I would just maybe suggest documenting which rows `unique[!](df, cols)` would retain. For `unique[!](df)` it's not relevant, but here it could be the source of confusion. IIRC it should be the first rows where specific combination of `cols` were encountered (so it matches the `nonunique(df, cols)`), but maybe we would like to avoid any commitments and state that this is undefined.

**Comment by nalimilan** [comment]
I think we c

File: src/abstractdataframe/abstractdataframe.jl [hunk 9]
```diff
@@ -614,7 +614,9 @@ unique!(df::AbstractDataFrame, cols)
 **Arguments**
 
 * `df` : the AbstractDataFrame
-* `cols` : the columns to compare – any valid DataFrames column specifier will work
+* `cols` :  column indicator (Symbol, Int, Vector{Symbol}, etc.)
+specifying the column(s) to compare. The first instance of each complete
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: That indication would make more sense in the "Result" section (for for the repeated requests...).

Comment (mkborregaard) [inline src/abstractdataframe/abstractdataframe.jl]: Good point. I have moved it down and tried to improve the wording, though it is still a bit wordy for my taste.

**Comment by nalimilan** [comment]
Thanks!

**Comment by alyst** [comment]
@mkborregaard Thanks for your patience! ;)

**Comment by mkborregaard** [comment]
Thanks a lot for your exact advice - I learned a lot from this!

**Comment by gtollefson** [comment]
Is there a way to select which duplicate row is selected by unique!(df, cols) by choosing the row with a greater value in another column? Something like, unique!(df, 2) but choose row with greater value in df[:3]?

**Comment by gtollefson** [comment]
I've received an answer via Alex Mellnik on the julialang slack channel. Something like this should work:
input = DataFrame(
    A = [1, 2, 2, 5, 5, 3],
    B = [1.2, 2.3, 1.1, 5.2, 2.2, 1.6],
    C = ["a", "b", "c", "d", "e", "f"]
)

by(df, :A) do df
    df[df[:B].==maximum(df[:B]), 2:end]
end

**Comment by nalimilan** [comment]
You can also sort the data frame first.

**Comment by gtollefson** [comment]
@nalimilan, thanks for you suggestion. I'm trying to sort columns 1:2 in ascending order but column 3 in descending order. With a matrix I would write something like matrix = sortrows(matrix, by=x->(x[1],x[2]), rev=false) and matrix = sortrows(matrix, by=x->(x[3]), rev=true) but prioritizing the sort by order of cols 1:2. Do you know of a way to do this in DataFrames or base Julia?

**Comment by gtollefson** [comment]
I've found the answer on another thread. `matrix = sortrows(matrix, by=x->(x[1],x[2],-x[3]))` works here.

https://stackoverflow.com/questions/35805099/julia-sorting-a-matrix-by-2-columns-in-different-orders
